File: ml/synthaea_ml/features/correlation.py
"""Extraction of the "multi-events per pid" feature vector, for a second ML scorer — a
complement to the cmdline vector (`features.py`), never merged with it into a single model
(decision of 2026-08-27, see docs/design/design-correlation-ml-vector.md).

**Must stay in sync with the Rust mirror** (crates/synthaea-ml/src/correlation_features.rs):
exact same definitions, same output order.

Expected input format: a list of JSON-Lines events, one object per event, as produced by
`edr-cli capture-events` / `edr-cli run` in `events.jsonl`:
    {"type": "exec",     "pid": 1234, "ts_ns": 1000000000, "comm": "curl", "cmdline": "..."}
    {"type": "connect",  "pid": 1234, "ts_ns": 2000000000, "daddr_v4": [127,0,0,1],
                         "dport": 4444, "is_ipv6": false}
    {"type": "fileopen", "pid": 1234, "ts_ns": 3000000000, "path": "...", "flags": 65}
    {"type": "network_flow", "pid": 1234, "ts_ns": 4000000000, "daddr_v4": [10,0,0,1],
                         "dport": 9999}

`network_flow` (netlink-observed, ADR-0008) is absorbed into the same `unique_daddr_count`/
`unique_dport_count` sets as `connect` — see `extract_features` below. `connect_count` stays
`connect`-only.

`flags` is logged raw; the filtering on write intent (access mode `O_WRONLY`/`O_RDWR`,
or `O_CREAT`) happens here, exactly mirroring `schema::has_write_intent` on the Rust side
(`features/behavior.py` applies the same test).
"""

# ── Rust mirror constants (POSIX flags) ───────────────────────────────────────

O_ACCMODE = 0o3
O_WRONLY = 0o1
O_RDWR = 0o2
O_CREAT = 0o100

FEATURE_NAMES = [
    "spawn_count",
    "connect_count",
    "filewrite_count",
    "unique_daddr_count",
    "unique_dport_count",
    "has_full_chain",
    "span_s",
    "event_count",
]


def _is_file_write(event: dict) -> bool:
    """Mirror of `schema::has_write_intent` (Rust, the ONE definition): a `fileopen`
    whose access mode is `O_WRONLY`/`O_RDWR`, or that carries `O_CREAT`. The access
    mode is a 2-bit *field* (`O_ACCMODE`), not independent bits — the invalid `0o3`
    combination is refused by the kernel (`EINVAL`), so it does not count as a write
    (this replaced a drifted bitmask test that disagreed with the Rust rules engine)."""
    if event.get("type") != "fileopen":
        return False
    flags = int(event.get("flags", 0))
    access_mode = flags & O_ACCMODE
    return access_mode in (O_WRONLY, O_RDWR) or bool(flags & O_CREAT)


def extract_features(events: list[dict], pid: int) -> list[float]:
    """`events`: all events of a correlation window (not only those of the requested pid —
    the pid filtering happens here, like `EventBus::events_for_pid` on the Rust side)."""
    pid_events = [e for e in events if e["pid"] == pid]

    spawn_count = sum(1 for e in pid_events if e["type"] == "exec")
    connect_count = sum(1 for e in pid_events if e["type"] == "connect")
    filewrite_count = sum(1 for e in pid_events if _is_file_write(e))

    # ADR-0008: Connect and NetworkFlow (netlink) are absorbed into the same
    # daddr/dport sets — one count of "distinct destinations touched" regardless of
    # which sensor captured them. connect_count above stays Connect-only.
    daddrs = {
        tuple(e["daddr_v4"]) for e in pid_events if e["type"] in ("connect", "network_flow")
    }
    dports = {e["dport"] for e in pid_events if e["type"] in ("connect", "network_flow")}

    has_full_chain = (
        1.0 if (spawn_count >= 1 and connect_count >= 1 and filewrite_count >= 1) else 0.0
    )

    if pid_events:
        span_s = (
            max(e["ts_ns"] for e in pid_events) - min(e["ts_ns"] for e in pid_events)
        ) / 1_000_000_000.0
    else:
        span_s = 0.0

    return [
        float(spawn_count),
        float(connect_count),
        float(filewrite_count),
        float(len(daddrs)),
        float(len(dports)),
        has_full_chain,
        span_s,
        float(len(pid_events)),
    ]


# ── Parity with the Rust tests (crates/synthaea-ml/src/correlation_features.rs) ──

if __name__ == "__main__":
    print("=== Tests correlation_features.py (mirror of the Rust tests) ===\n")

    # vecteur_vide_pour_pid_absent
    assert extract_features([], 1234) == [0.0] * 8
    assert extract_features([{"type": "exec", "pid": 1, "ts_ns": 0}], 1234) == [0.0] * 8
    print("[OK] absent pid → zero vector")

    # chaine_complete_a_has_full_chain_a_un
    evs = [
        {"type": "exec", "pid": 99, "ts_ns": 0},
        {
            "type": "connect",
            "pid": 99,
            "ts_ns": 1_000_000_000,
            "daddr_v4": [127, 0, 0, 1],
            "dport": 4444,
        },
        {"type": "fileopen", "pid": 99, "ts_ns": 2_000_000_000, "flags": O_WRONLY | O_CREAT},
    ]
    f = extract_features(evs, 99)
    assert f[0] == 1.0 and f[1] == 1.0 and f[2] == 1.0, f
    assert f[5] == 1.0, "has_full_chain"
    assert f[6] == 2.0, "span_s 0->2s"
    assert f[7] == 3.0, "event_count"
    print(f"[OK] full chain: {f}")

    # spawn_seul_n_a_pas_has_full_chain
    assert extract_features([{"type": "exec", "pid": 1, "ts_ns": 0}], 1)[5] == 0.0
    print("[OK] spawn alone → has_full_chain=0")

    # a read-only fileopen does not count as a filewrite
    ro = [{"type": "fileopen", "pid": 5, "ts_ns": 0, "flags": 0}]  # O_RDONLY
    assert extract_features(ro, 5)[2] == 0.0, "read-only ≠ filewrite"
    print("[OK] fileopen O_RDONLY → filewrite_count=0")

    # destinations_multiples_comptees_une_fois_chacune
    net = [
        {"type": "connect", "pid": 7, "ts_ns": 0, "daddr_v4": [127, 0, 0, 1], "dport": 4444},
        {
            "type": "connect",
            "pid": 7,
            "ts_ns": 1_000_000_000,
            "daddr_v4": [127, 0, 0, 1],
            "dport": 4444,
        },
        {
            "type": "connect",
            "pid": 7,
            "ts_ns": 2_000_000_000,
            "daddr_v4": [127, 0, 0, 1],
            "dport": 8080,
        },
    ]
    fn = extract_features(net, 7)
    assert fn[1] == 3.0, "connect_count"
    assert fn[3] == 1.0, "unique_daddr_count"
    assert fn[4] == 2.0, "unique_dport_count"
    print(f"[OK] distinct destinations: {fn}")

    # network_flow_absorbe_dans_les_memes_sets_daddr_dport_que_connect (ADR-0008)
    mixte = [
        {
            "type": "connect",
            "pid": 7,
            "ts_ns": 0,
            "daddr_v4": [127, 0, 0, 1],
            "dport": 4444,
        },
        {
            "type": "network_flow",
            "pid": 7,
            "ts_ns": 1_000_000_000,
            "daddr_v4": [10, 0, 0, 1],
            "dport": 4444,
        },
        {
            "type": "network_flow",
            "pid": 7,
            "ts_ns": 2_000_000_000,
            "daddr_v4": [10, 0, 0, 2],
            "dport": 9999,
        },
    ]
    fm = extract_features(mixte, 7)
    assert fm[3] == 3.0, "unique_daddr_count spans connect + network_flow"
    assert fm[4] == 2.0, "unique_dport_count spans connect + network_flow"
    assert fm[7] == 3.0, "event_count includes network_flow"
    print(f"[OK] network_flow absorbed into connect's daddr/dport sets: {fm}")

    # pids_differents_isoles
    mix = [
        {"type": "exec", "pid": 1, "ts_ns": 0},
        {
            "type": "connect",
            "pid": 2,
            "ts_ns": 1_000_000_000,
            "daddr_v4": [1, 1, 1, 1],
            "dport": 4444,
        },
    ]
    assert extract_features(mix, 1)[1] == 0.0
    assert extract_features(mix, 2)[0] == 0.0
    print("[OK] isolated pids")

    print("\nAll tests pass.")

File: crates/ml/src/features/correlation.rs
//! Correlation features for the T2 behaviour scorer — the "multi-events per pid"
//! vector, Rust mirror of `ml/synthaea_ml/features/correlation.py`.
//!
//! A second, complementary signal to [`super::cmdline`], never merged into one model
//! (2026-08-27 decision): the two vectors differ in temporal availability (a cmdline
//! is ready the instant an `Exec` arrives; this one only once the correlator window
//! is populated) and in statistical nature. Parity with the Python definition is
//! pinned end-to-end by `crates/ml/tests/capture_parity.rs` (a real `events.jsonl`
//! capture → these vectors, checked against the Python aggregator's output).
//!
//! No trained model ships against this yet: it needs a dedicated benign multi-event
//! capture campaign (#44), not the isolated-cmdline baselines we have. The extractor
//! and its parity are migrated now so the training data can be scored the moment it
//! exists.

use std::collections::HashSet;

use correlator::EventBus;
use schema::Event;

/// Write intent on a `FileOpen` — mirror of `correlator`'s crate-private
/// `is_file_write` and of `correlation.py::_is_file_write`. Duplicated deliberately:
/// the correlator does not export it, and a bit test is cheaper to mirror than to
/// plumb a new public API through the platform boundary.
fn is_file_write(event: &Event) -> bool {
    const O_WRONLY: u32 = 0o1;
    const O_RDWR: u32 = 0o2;
    const O_CREAT: u32 = 0o100;
    match event {
        Event::FileOpen(f) => f.flags & (O_WRONLY | O_RDWR | O_CREAT) != 0,
        _ => false,
    }
}

/// Feature names in output order — must match `correlation.py::FEATURE_NAMES`.
pub const FEATURE_NAMES: [&str; 8] = [
    "spawn_count",
    "connect_count",
    "filewrite_count",
    "unique_daddr_count",
    "unique_dport_count",
    "has_full_chain",
    "span_s",
    "event_count",
];

/// Event variants this vector is built from — process spawn, outbound connect, file
/// open, netlink-observed flow (ADR-0008). `span_s` and `event_count` are computed
/// over these only: the Python aggregator (`aggregate_correlation._flatten`,
/// `RAW_EVENT_TYPES`) drops every other category before windowing, so counting e.g. a
/// `DnsQuery` here would silently inflate both features relative to the training
/// vectors.
///
/// ADR-0008 side effect: gating the whole vector on this filter means `event_count`
/// and `span_s` grow for any process with netlink traffic, not just the two fields
/// `NetworkFlow` feeds below. `connect_count` (T1's `BehaviorVector`) is unaffected —
/// it filters on `Event::Connect` explicitly, independent of this gate.
fn is_modeled(event: &Event) -> bool {
    matches!(
        event,
        Event::Exec(_) | Event::Connect(_) | Event::FileOpen(_) | Event::NetworkFlow(_)
    )
}

/// The 8-feature correlation vector for `pid` over the bus's current window, in
/// [`FEATURE_NAMES`] order. Pid filtering happens here (mirror of the Python side
/// taking the whole window), via [`EventBus::events_for_pid`].
#[must_use]
pub fn extract_features(bus: &EventBus, pid: u32) -> [f32; 8] {
    let events: Vec<&Event> = bus.events_for_pid(pid).filter(|e| is_modeled(e)).collect();

    let spawn_count = events
        .iter()
        .filter(|e| matches!(e, Event::Exec(_)))
        .count();
    let connect_count = events
        .iter()
        .filter(|e| matches!(e, Event::Connect(_)))
        .count();
    let filewrite_count = events.iter().filter(|e| is_file_write(e)).count();

    // ADR-0008: NetworkFlow (netlink) and Connect are absorbed into the same
    // daddr/dport sets — one count of "distinct destinations touched", regardless of
    // which sensor captured them (mirrors BEACON's check_beacon/check_beacon_flow
    // convergence on shared state).
    let mut daddrs = HashSet::new();
    let mut dports = HashSet::new();
    for e in &events {
        match e {
            Event::Connect(c) => {
                daddrs.insert(c.daddr);
                dports.insert(c.dport);
            }
            Event::NetworkFlow(f) => {
                daddrs.insert(f.daddr);
                dports.insert(f.dport);
            }
            _ => {}
        }
    }

    let has_full_chain = if spawn_count >= 1 && connect_count >= 1 && filewrite_count >= 1 {
        1.0
    } else {
        0.0
    };

    let span_s = match (
        events.iter().map(|e| e.meta().timestamp_ns).min(),
        events.iter().map(|e| e.meta().timestamp_ns).max(),
    ) {
        (Some(min), Some(max)) => (max - min) as f32 / 1_000_000_000.0,
        _ => 0.0,
    };

    [
        spawn_count as f32,
        connect_count as f32,
        filewrite_count as f32,
        daddrs.len() as f32,
        dports.len() as f32,
        has_full_chain,
        span_s,
        events.len() as f32,
    ]
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use schema::{
        ConnectEvent, DnsQueryEvent, EventMeta, ExecEvent, FileOpenEvent, NetworkFlowEvent, User,
    };

    use super::*;

    fn meta(pid: u32, ts_ns: u64) -> EventMeta {
        EventMeta {
            pid,
            user: User::Unix { uid: 0, gid: 0 },
            timestamp_ns: ts_ns,
            comm: "proc".into(),
            ..schema::fixtures::meta()
        }
    }

    fn exec(pid: u32, ts_ns: u64) -> Event {
        Event::Exec(ExecEvent {
            meta: meta(pid, ts_ns),
            ..schema::fixtures::exec()
        })
    }

    fn connect(pid: u32, ts_ns: u64, dport: u16) -> Event {
        Event::Connect(ConnectEvent {
            meta: meta(pid, ts_ns),
            daddr: "127.0.0.1".parse().unwrap(),
            dport,
        })
    }

    fn network_flow(pid: u32, ts_ns: u64, daddr: &str, dport: u16) -> Event {
        Event::NetworkFlow(NetworkFlowEvent {
            meta: meta(pid, ts_ns),
            local_port: 54321,
            daddr: daddr.parse().unwrap(),
            dport,
            protocol: 6,
            ..schema::fixtures::network_flow()
        })
    }

    fn file_write(pid: u32, ts_ns: u64) -> Event {
        Event::FileOpen(FileOpenEvent {
            meta: meta(pid, ts_ns),
            path: String::new(),
            flags: 0o1 | 0o100, // O_WRONLY | O_CREAT
        })
    }

    #[test]
    fn empty_vector_for_absent_pid() {
        let bus = EventBus::new(Duration::from_secs(60));
        assert_eq!(extract_features(&bus, 1234), [0.0; 8]);
    }

    #[test]
    fn full_chain_sets_has_full_chain() {
        let mut bus = EventBus::new(Duration::from_secs(60));
        bus.push(exec(99, 0));
        bus.push(connect(99, 1_000_000_000, 4444));
        bus.push(file_write(99, 2_000_000_000));

        let f = extract_features(&bus, 99);
        assert_eq!(f[0], 1.0, "spawn_count");
        assert_eq!(f[1], 1.0, "connect_count");
        assert_eq!(f[2], 1.0, "filewrite_count");
        assert_eq!(f[5], 1.0, "has_full_chain");
        assert_eq!(f[6], 2.0, "span_s");
        assert_eq!(f[7], 3.0, "event_count");
    }

    #[test]
    fn unique_destinations_counted_once() {
        let mut bus = EventBus::new(Duration::from_secs(60));
        bus.push(connect(7, 0, 4444));
        bus.push(connect(7, 1_000_000_000, 4444)); // same dest
        bus.push(connect(7, 2_000_000_000, 8080)); // different port

        let f = extract_features(&bus, 7);
        assert_eq!(f[1], 3.0, "connect_count");
        assert_eq!(f[3], 1.0, "unique_daddr_count");
        assert_eq!(f[4], 2.0, "unique_dport_count");
    }

    #[test]
    fn network_flow_absorbed_into_same_daddr_dport_sets_as_connect() {
        // ADR-0008: Connect and NetworkFlow share one "distinct destinations" count —
        // an attacker whose traffic is seen only by netlink still shows up here.
        let mut bus = EventBus::new(Duration::from_secs(60));
        bus.push(connect(7, 0, 4444));
        bus.push(network_flow(7, 1_000_000_000, "10.0.0.1", 4444)); // same dest as connect
        bus.push(network_flow(7, 2_000_000_000, "10.0.0.2", 9999)); // netlink-only dest

        let f = extract_features(&bus, 7);
        assert_eq!(f[3], 3.0, "unique_daddr_count spans Connect + NetworkFlow");
        assert_eq!(f[4], 2.0, "unique_dport_count spans Connect + NetworkFlow");
        assert_eq!(
            f[7], 3.0,
            "event_count includes NetworkFlow (is_modeled gate)"
        );
    }

    #[test]
    fn non_modeled_events_do_not_perturb_span_or_count() {
        // The Python aggregator drops DnsQuery et al. before windowing; span_s and
        // event_count here must match — a later, unrelated DNS lookup must not
        // stretch the span or bump the count.
        let mut bus = EventBus::new(Duration::from_secs(60));
        bus.push(exec(5, 0));
        bus.push(connect(5, 1_000_000_000, 443));
        bus.push(Event::DnsQuery(DnsQueryEvent {
            meta: meta(5, 9_000_000_000),
            query: "c2.test".into(),
            qtype: 1,
            result: None,
            status: 0,
        }));

        let f = extract_features(&bus, 5);
        assert_eq!(f[7], 2.0, "event_count counts only modeled events");
        assert_eq!(f[6], 1.0, "span_s ignores the later DNS lookup");
    }

    #[test]
    fn pids_are_isolated() {
        let mut bus = EventBus::new(Duration::from_secs(60));
        bus.push(exec(1, 0));
        bus.push(connect(2, 1_000_000_000, 4444));
        assert_eq!(extract_features(&bus, 1)[1], 0.0);
        assert_eq!(extract_features(&bus, 2)[0], 0.0);
    }
}
